### DISCUSSION TITLE
send_message not working after upgrade to latest fcm-django 10.0.12

### DISCUSSION THREAD

**Discussion by quirinhamp** [xtrinch/fcm-django#249] [Q&A]
**Error:**
Django Version: | 4.2.2
-- | --
ValueError
Exactly one of token, topic or condition must be specified.
/home/ubuntu/comfortid_backend/venv/lib/python3.10/site-packages/firebase_admin/_messaging_encoder.py, line 681, in default
dsm.views.overview
/home/ubuntu/.local/bin/uwsgi
3.10.6

**settings.py**
from firebase_admin import initialize_app, credentials
cred = credentials.Certificate(cred_path)
initialize_app(cred)

**view.py**
  android_devices = devices.filter(type='android')
  obj = )
  message = Message(notification=Notification(title=form.cleaned_data['message_title'],
                                       body=form.cleaned_data['message'],
                                       image=None)
                                       ,
                                        data=NotificationSerializer(not_obj).data,
                                        topic="DSM"
                                        )

  android_devices.send_message(message)

**admin**

FCM device | Device ID | Name | Type | User | Is active | Creation date
-- | -- | -- | -- | -- | -- | --
  | 93bf7f511865531ed9cdb2bd0bf6f3a3960e22052b58390b202486ac |   | ios | admin@comfortid.org |   | 30. Juni 2023 16:10
  | 37adcc81dcaba724 |   | android | admin@comfrtid.org |   | 30. Juni 2023 14:04

FCM device
	
[Device ID](https://comfortid.org/admin/fcm_django/fcmdevice/?o=2)
	
[Name](https://comfortid.org/admin/fcm_django/fcmdevice/?o=3)
	
[Type](https://comfortid.org/admin/fcm_django/fcmdevice/?o=4)
	
[User](https://comfortid.org/admin/fcm_django/fcmdevice/?o=5)
	
[Is active](https://comfortid.org/admin/fcm_django/fcmdevice/?o=6)
	
[Creation date](https://comfortid.org/admin/fcm_django/fcmdevice/?o=7)
	[93bf7f511865bd0bf6f3a3960e22052b58390b202486ac](https://comfortid.org/admin/fcm_django/fcmdevice/2/change/)	93bf7f511865531ed9ccf35da90db2bd0bf6f3a3960e22052b58390b202486ac		ios	admin@cortid.org	True	30. Juni 2023 16:10
	[37adcc81d24](https://comfortid.org/admin/fcm_django/fcmdevice/1/change/)	37adcca724		android	admin@comforid.org	True	30. Juni 2023 14:04

What is wrong? Please help! In the past, no token was required for bulk send_message. The devices are created according to requirements.

**Comment by xtrinch** [comment]
@quirinhamp Can you confirm none of your device FCM tokens in database are empty/null?

**Comment by quirinhamp** [comment]
Yes I can, device_id and registration_token are not null
also updated to latest version and I have the same behavior.

**Reply by quirinhamp** [reply]
The issue is that one can not declare a topic in Message since fcm-django adds token:

in line 679 of  _messaging_encoder.py:

` target_count = sum([t in result for t in ['token', 'topic', 'condition']])
        if target_count != 1:
            raise ValueError('Exactly one of token, topic or condition must be specified.')`

**Reply by xtrinch** [reply]
Try using send_topic_message for topic messages

**Comment by xtrinch** [comment]
Thanks for reporting this. The underlying limitation here is Firebase topic targeting: topic sends are separate from device/queryset sends, so `send_message()` is not the right API for sending to a topic. The README still showed a misleading example, so I opened PR #300 to correct the docs and point topic sends to `FCMDevice.send_topic_message(...)`: https://github.com/xtrinch/fcm-django/pull/300
